File: services/auth.service.js
/**
 * BarberEase – Auth Service
 * 
 * Business logic for authentication:
 *   - Customer registration
 *   - Customer login
 *   - Admin login
 *   - Password hashing and verification
 *   - JWT token generation
 *   - Password reset
 */

const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const config = require('../config/app');
const { Customer, Admin } = require('../models');

/**
 * Register a new customer account.
 * Hashes password, creates record, returns JWT.
 */
async function registerCustomer(data) {
  // Check for existing email
  const existing = await Customer.scope('withPassword').findOne({
    where: { email: data.email }
  });

  if (existing) {
    const error = new Error('Email already registered');
    error.statusCode = 409;
    throw error;
  }

  // Hash password
  const hashedPassword = await bcrypt.hash(data.password, config.bcrypt.saltRounds);

  // Create customer
  const customer = await Customer.create({
    name: data.name,
    email: data.email,
    phone: data.phone,
    password: hashedPassword,
    address: data.address || null,
    gender: data.gender || null
  });

  // Generate JWT
  const token = generateJWT({
    id: customer.id,
    email: customer.email,
    role: 'customer',
    type: 'customer'
  });

  // Return customer without password
  const customerData = await Customer.findByPk(customer.id);

  return { customer: customerData, token };
}

/**
 * Authenticate a customer with email and password.
 */
async function loginCustomer(email, password) {
  // Find customer with password included
  const customer = await Customer.scope('withPassword').findOne({
    where: { email }
  });

  if (!customer) {
    const error = new Error('Invalid email or password');
    error.statusCode = 401;
    throw error;
  }

  if (!customer.is_active) {
    const error = new Error('Account is deactivated. Contact support.');
    error.statusCode = 403;
    throw error;
  }

  // Compare password
  const isMatch = await bcrypt.compare(password, customer.password);

  if (!isMatch) {
    const error = new Error('Invalid email or password');
    error.statusCode = 401;
    throw error;
  }

  // Update last login
  await customer.update({ last_login: new Date() });

  // Generate JWT
  const token = generateJWT({
    id: customer.id,
    email: customer.email,
    role: 'customer',
    type: 'customer'
  });

  // Return customer without password
  const customerData = await Customer.findByPk(customer.id);

  return { customer: customerData, token };
}

/**
 * Authenticate an admin with username and password.
 */
async function loginAdmin(username, password) {
  const admin = await Admin.scope('withPassword').findOne({
    where: { username }
  });

  if (!admin) {
    const error = new Error('Invalid username or password');
    error.statusCode = 401;
    throw error;
  }

  if (!admin.is_active) {
    const error = new Error('Admin account is deactivated.');
    error.statusCode = 403;
    throw error;
  }

  const isMatch = await bcrypt.compare(password, admin.password);

  if (!isMatch) {
    const error = new Error('Invalid username or password');
    error.statusCode = 401;
    throw error;
  }

  // Update last login
  await admin.update({ last_login: new Date() });

  const token = generateJWT({
    id: admin.id,
    email: admin.email,
    role: admin.role,
    type: 'admin'
  });

  const adminData = await Admin.findByPk(admin.id);

  return { admin: adminData, token };
}

/**
 * Reset customer password (simplified — no email verification).
 * In production, implement email-based token verification.
 */
async function resetPassword(email, newPassword) {
  const customer = await Customer.scope('withPassword').findOne({
    where: { email }
  });

  if (!customer) {
    const error = new Error('No account found with this email');
    error.statusCode = 404;
    throw error;
  }

  const hashedPassword = await bcrypt.hash(newPassword, config.bcrypt.saltRounds);
  await customer.update({ password: hashedPassword });

  return { message: 'Password reset successfully' };
}

/**
 * Generate a JWT token from the user payload.
 */
function generateJWT(payload) {
  return jwt.sign(payload, config.jwt.secret, {
    expiresIn: config.jwt.expiresIn
  });
}

module.exports = {
  registerCustomer,
  loginCustomer,
  loginAdmin,
  resetPassword
};

File: src/main/java/com/barberease/services/AuthService.java
package com.barberease.services;

import com.barberease.models.Admin;
import com.barberease.models.Customer;
import com.barberease.repositories.AdminRepository;
import com.barberease.repositories.CustomerRepository;
import com.barberease.security.JwtUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
@Transactional
public class AuthService {

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private AdminRepository adminRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtils jwtUtils;

    public Map<String, Object> registerCustomer(Customer customerData) {
        Optional<Customer> existing = customerRepository.findByEmail(customerData.getEmail());
        if (existing.isPresent()) {
            throw new IllegalArgumentException("Email already registered");
        }

        customerData.setPassword(passwordEncoder.encode(customerData.getPassword()));
        Customer saved = customerRepository.save(customerData);

        String token = jwtUtils.generateToken(saved.getId(), saved.getEmail(), "customer", "customer");

        Map<String, Object> result = new HashMap<>();
        result.put("customer", saved);
        result.put("token", token);
        return result;
    }

    public Map<String, Object> loginCustomer(String email, String password) {
        Customer customer = customerRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Invalid email or password"));

        if (!customer.isActive()) {
            throw new IllegalStateException("Account is deactivated. Contact support.");
        }

        if (!passwordEncoder.matches(password, customer.getPassword())) {
            throw new IllegalArgumentException("Invalid email or password");
        }

        customer.setLastLogin(LocalDateTime.now());
        customerRepository.save(customer);

        String token = jwtUtils.generateToken(customer.getId(), customer.getEmail(), "customer", "customer");

        Map<String, Object> result = new HashMap<>();
        result.put("customer", customer);
        result.put("token", token);
        return result;
    }

    public Map<String, Object> loginAdmin(String username, String password) {
        Admin admin = adminRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("Invalid username or password"));

        if (!admin.isActive()) {
            throw new IllegalStateException("Admin account is deactivated.");
        }

        if (!passwordEncoder.matches(password, admin.getPassword())) {
            throw new IllegalArgumentException("Invalid username or password");
        }

        admin.setLastLogin(LocalDateTime.now());
        adminRepository.save(admin);

        String token = jwtUtils.generateToken(admin.getId(), admin.getEmail(), admin.getRole(), "admin");

        Map<String, Object> result = new HashMap<>();
        result.put("admin", admin);
        result.put("token", token);
        return result;
    }

    public Map<String, Object> resetPassword(String email, String newPassword) {
        Customer customer = customerRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("No account found with this email"));

        customer.setPassword(passwordEncoder.encode(newPassword));
        customerRepository.save(customer);

        Map<String, Object> result = new HashMap<>();
        result.put("message", "Password reset successfully");
        return result;
    }
}
